WIP: promql: Implement support for versioned OTel semantic conventions

aknuds1 opened the pull request:
<!--
    - Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

#### Which issue(s) does the PR fix:
<!--
If it applies.
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
More at https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
Implement support for versioned OTel semantic conventions, with automatic querying for all metric versions. Behind feature flag "semconv-versioned-read" - this is a working title, final name is TODO :)

TODO:

- [ ] Transform the label names of all series belonging to other versions to correspond to the queried version
- [ ] Compute semantic conventions group IDs (how?) - potentially unnecessary with next version of semantic conventions format
- [ ] Use OTel schema format v2, where definitions are included (@jsuereth knows)
- [ ] Decide on feature flag name
- [ ] Resolve TODOs in code

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-notes block below.
Otherwise, please describe what should be mentioned in the CHANGELOG. Use the following prefixes:
[FEATURE] [ENHANCEMENT] [PERF] [BUGFIX] [SECURITY] [CHANGE]
Refer to the existing CHANGELOG for inspiration:  https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md
If you need help formulating your entries, consult the reviewer(s).
-->
```release-notes
[FEATURE] Add support for versioned OTel semantic conventions, with automatic querying for all metric versions.
```

nicolastakashi commented on storage/semconv/engine.go:
```diff
@@ -0,0 +1,368 @@
+	}
+
+	// Fetch semantic conventions for metric metadata.
+	sc, ok := e.semconvCache.get(semconvURL)
```
I'm wondering if we should have something like e.semconv.get(semconvURL) and the get method does all the orchestration between cache and remote server.

aknuds1 replied:
Quite possible. I'll have to look the code over again when time permits (next week?).

aknuds1 committed:
semconv: Add versioned OTel semantic-conventions read

Add storage/semconv, a query-time wrapper around storage.Storage that
recognises two reserved matchers in PromQL queries — __semconv_url__ and
__schema_url__ — and fans the query out across the historical metric and
attribute names declared by the referenced OTel schema, merging the
results under the requested version's naming.

The two matchers are:
- __semconv_url__ selects a semantic-conventions version from the embedded
  registry and supplies the metric metadata __schema_url__ needs; on its
  own it has no effect.
- __schema_url__ selects an OTel schema file and triggers schema-version
  rename fan-out (e.g. test.counter -> test) from the schema's versions
  section, recovering a metric's historical names.

The package ships an embedded registry of semconv/OTel schema files under
storage/semconv/registry/. Arbitrary HTTP URLs and local filesystem paths
are rejected to keep the matchers from acting as a server-side fetch
primitive.

The experimental --enable-feature=semconv-versioned-read flag opts in to
wrapping the storage; off by default.

A self-contained demo under documentation/examples/semconv-translation/
illustrates a native-OTel producer whose metric is renamed across semconv
versions (test.counter in 1.0.0 -> test in 1.1.0), showing how
__semconv_url__ with __schema_url__ maintains query continuity across the
rename.

Usage:
  cd documentation/examples/semconv-translation
  ./demo.sh [port]
Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
semconv: Add OTLP translation-strategy fan-out

Extend storage/semconv with a third reserved matcher, __otlp_strategy__,
an axis orthogonal to __schema_url__: it triggers fan-out across the OTLP
metric and attribute translation strategies (UnderscoreEscapingWithSuffixes,
UnderscoreEscapingWithoutSuffixes, NoUTF8EscapingWithSuffixes, NoTranslation)
and selects the dialect results are rendered in.

The queried metric and label names must be written in the chosen dialect;
they are reverse-resolved to the canonical OTel names before fan-out, and
the merged results are re-rendered in that dialect. A label matcher that
targets a known attribute in a different escaping is applied verbatim and
surfaces a warning.

The demo under documentation/examples/semconv-translation/ is expanded to a
three-era producer migration (semconv 1.0.0 escaped, 1.1.0 escaped, and
1.1.0 native), showing how __otlp_strategy__ maintains query continuity
across an OTLP-strategy migration and how adding __schema_url__ additionally
recovers the older semconv 1.0.0 era.

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
Closing as superseded by #18858.
